File: 047_Permutation_II.py
# Given a collection of numbers that might contain duplicates,
# return all possible unique permutations.

# For example,
# [1,1,2] have the following unique permutations:
# [
#   [1,1,2],
#   [1,2,1],
#   [2,1,1]
# ]
# Subscribe to see which companies asked this question


# Runtime: 152ms
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def helper(nums, length, used, ans):
            if length == len(nums):
                res.append(ans)
                return
            this_used = {num: 0 for num in nums}
            for i, num in enumerate(nums):
                if used[i] == 0 and this_used[num]==0:
                    this_used[num]=1
                    used[i] = 1
                    helper(nums, length+1, used, ans+[num])
                    used[i] = 0


        res = []
        used = {i:0 for i, _ in enumerate(nums)}
        helper(nums, 0, used, [])
        return res

if __name__ == '__main__':
    sol = Solution()
    print sol.permuteUnique([1,2])

File: java/047_Permutation_II.java
// Given a collection of numbers that might contain duplicates,
// return all possible unique permutations.

// For example,
// [1,1,2] have the following unique permutations:
// [
//   [1,1,2],
//   [1,2,1],
//   [2,1,1]
// ]

// Runtime: 21 ms



public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        Integer[] current = new Integer[nums.length];
        boolean[] chosen = new boolean[nums.length];
        helper(res, nums, current, 0, chosen);
        return res;
    }
    
    private void helper(List<List<Integer>> res, int[] nums, Integer[] current, int pos, boolean[] chosen) {
        if (pos == nums.length) {
            res.add(new ArrayList<>(Arrays.asList(current)));
            return;
        }
        Set<Integer> visited = new HashSet<>();
        for(int i = 0; i < nums.length; i++) {
            if (!visited.contains(nums[i]) && !chosen[i]) {
                visited.add(nums[i]);
                chosen[i] = true;
                current[pos] = nums[i];
                helper(res, nums, current, pos + 1, chosen);
                chosen[i] = false;
            }
        }
    }
}